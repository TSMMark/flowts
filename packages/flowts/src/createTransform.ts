import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from '@flowts/babel-plugin-recast';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      compact: false,
      babelrc: false,
      configFile: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'babel-ts';
    return prettier.format(transformResult.code as string, {
      ...prettierConfig,
      // avoid changing template strings
      embeddedLanguageFormatting: 'off',
    });
  };
}
