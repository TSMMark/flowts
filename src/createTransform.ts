import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from './recastPlugin';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      babelrc: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'typescript';
    return prettier.format(transformResult.code as string, prettierConfig);
  };
}
