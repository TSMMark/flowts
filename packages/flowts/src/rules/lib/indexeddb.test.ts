import plugin from '../../tsTypesPlugin';
import { createTransform } from '../../createTransform';
const transform = createTransform([plugin]);

describe('lib/indexeddb.js', () => {
  describe('globals', () => {
    describe('IDBEnvironment', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBEnvironment {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBFactory', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBFactory {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBRequest', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBRequest {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBOpenDBRequest', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBOpenDBRequest {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBDatabase', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBDatabase {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBTransaction', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBTransaction {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBObjectStore', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBObjectStore {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBIndex', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBIndex {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBKeyRange', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBKeyRange {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBCursor', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBCursor {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('IDBCursorWithValue', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 implements IDBCursorWithValue {};
          `)
        ).toMatchSnapshot();
      });
    });
  });

  describe('modules', () => {
    test('has no modules', () => {});
  });
});
