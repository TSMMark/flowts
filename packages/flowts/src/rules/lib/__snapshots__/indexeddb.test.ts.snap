// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/indexeddb.js globals IDBCursor generated - interface 1`] = `
"class A1 implements IDBCursor {}
"
`;

exports[`lib/indexeddb.js globals IDBCursor generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBCursorWithValue generated - interface 1`] = `
"class A1 implements IDBCursorWithValue {}
"
`;

exports[`lib/indexeddb.js globals IDBCursorWithValue generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBDatabase generated - interface 1`] = `
"class A1 implements IDBDatabase {}
"
`;

exports[`lib/indexeddb.js globals IDBDatabase generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBEnvironment generated - interface 1`] = `
"class A1 implements IDBEnvironment {}
"
`;

exports[`lib/indexeddb.js globals IDBEnvironment generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBFactory generated - interface 1`] = `
"class A1 implements IDBFactory {}
"
`;

exports[`lib/indexeddb.js globals IDBFactory generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBIndex generated - interface 1`] = `
"class A1 implements IDBIndex {}
"
`;

exports[`lib/indexeddb.js globals IDBIndex generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBKeyRange generated - interface 1`] = `
"class A1 implements IDBKeyRange {}
"
`;

exports[`lib/indexeddb.js globals IDBKeyRange generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBObjectStore generated - interface 1`] = `
"class A1 implements IDBObjectStore {}
"
`;

exports[`lib/indexeddb.js globals IDBObjectStore generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBOpenDBRequest generated - interface 1`] = `
"class A1 implements IDBOpenDBRequest {}
"
`;

exports[`lib/indexeddb.js globals IDBOpenDBRequest generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBRequest generated - interface 1`] = `
"class A1 implements IDBRequest {}
"
`;

exports[`lib/indexeddb.js globals IDBRequest generated - type 1`] = `""`;

exports[`lib/indexeddb.js globals IDBTransaction generated - interface 1`] = `
"class A1 implements IDBTransaction {}
"
`;

exports[`lib/indexeddb.js globals IDBTransaction generated - type 1`] = `""`;
